### DISCUSSION TITLE
İsim uzayında kayboldum

### DISCUSSION THREAD

**Discussion by BunyaminAkdeniz** [19/classroom#6] [Ruby] [answered]
Sorum şudur
```ruby
VAL = 'Global'
 
module Foo
  VAL = 'Foo Local'
 
  class Bar
    def value1
      VAL
    end
  end
end
 
class Foo::Bar
  def value2
    VAL
  end
end

```

Foo::Bar.new.value1   #  => "Foo Local"
Foo::Bar.new.value2   #  =>  "Global"

ama neden ?

**Comment by roktas** [comment]
Kapsam (scope) veya daha teknik ifadeyle "Sözdizimsel kapsam (lexical scoping)" kurallarından dolayı...  `module Foo; end` bir kapsam açıyor ve (sonra) kapıyor.  `Bar` sınıfındaki `value1` metodu kapsamdaki en yakın sabite erişir ki o da `'Foo Local'`.  Tahmin edeceğin gibi `class Foo::Bar; end` de yeni bir kapsam açıyor, `value2` `VAL` aradığında bu yeni kapsamda bulamıyor ve bir üste çıkıyor; ki orada da `'Global'` olarak buluyor.

**Comment by roktas** [comment, marked as answer]
İşin sırrı `class Foo::Bar` demen.  Bu nedenle `VAL` `class Foo::Bar`'ı çevreleyen kapsama atlar.  Halbuki şöyle dersen `VAL` `Foo` modülünde bulunur:

```ruby
VAL = 'Global'
 
module Foo
  VAL = 'Foo Local'
 
  class Bar
    def value1
      VAL
    end
  end
end
 
module Foo
  class Bar
    def value2
      VAL
    end
  end
end

Foo::Bar.new.value1 # => "Foo Local"
Foo::Bar.new.value2 # => "Foo Local"
```

Ruby'nin kapsam davranışı böyle...

**Comment by roktas** [comment]
İlave not.  `Foo::Bar` kompakt söz diziminin kullanımı biraz da bu tür nedenlerle önerilmiyor.  Rubocop uyarısı:

![image](https://user-images.githubusercontent.com/47842/150700806-b11b5291-43aa-4a6c-862d-396946dc053d.png)

Evet, bu gösterim seni bir seviye girintiden kurtarıyor ama kapsam kurallarının işletilmesinde böyle sürpriz bir yan etki ortaya çıkıyor.  Kıssadan hisse Rubocop'un sözünü daima dinleyin.

**Reply by BunyaminAkdeniz** [reply]
Çok teşekkür ederim hocam

**Comment by roktas** [comment]
Rubocop'un sözüne uy ve kodu önerdiği şekle getir.  Ama kompakt kullanım bir nedenle gerekiyorsa (ör. bir başkasının yazdığı koda katkıda bulunuyorsun), bu durumda doğru davranış için şöyle yapacaksın:

```ruby
VAL = 'Global'

module Foo
  VAL = 'Foo Local'

  class Bar
    def value1
      VAL
    end
  end
end

class Foo::Bar
  def value2
    Foo::VAL
  end
end

```

**Reply by BunyaminAkdeniz** [reply]
Ruby'nin Sözdizimsel kapsam (lexical scoping)" kurallarını net anlamak küçük testler yapıyordum. Çok net anladım hocam sağ olun.
